backend/src/v2v3/dex/data.ts

```diff
@@ -0,0 +1,182 @@
+import { bqClient } from "../../bigQuery";
+import { redisClient, getOrThrow } from "../../redisSetup";
+import { getErrorMessage, getPrice } from "./utils";
+
+const bigQueryEndpointBase = "crypto-stellar.crypto_stellar_2";
+
+async function fetchBigQueryData(query: string) {
+  try {
+    const [job] = await bqClient.createQueryJob(query);
+    console.log("running bq query:", query);
+    const [result] = await job.getQueryResults();
+    return result;
+  } catch (error: unknown) {
+    const message = getErrorMessage(error);
+    console.error("QUERY ERROR: ", message);
+    return null;
+  }
+}
+
+async function fetchMemoizedData(key: string, query: string) {
+  try {
+    const cachedData = await getOrThrow(redisClient, key);
+    const output = JSON.parse(cachedData);
+    return output;
+  } catch (error: unknown) {
+    const message = getErrorMessage(error);
+    if (message == "redis key not found") {
+      const output = await fetchBigQueryData(query);
+      await redisClient.set(key, JSON.stringify(output));
+      return output;
+    }
+    console.error("ERROR: ", message);
+  }
+}
+
+export async function getPaymentsData() {
+  const query = `
+    SELECT count(*) as data 
+    FROM ${bigQueryEndpointBase}.enriched_history_operations 
+    WHERE type in (1, 2, 13) 
+    AND closed_at >= TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL -24 HOUR) 
+    AND successful is true
+  `;
+  const payments = await fetchMemoizedData("dex-payments24h", query);
+
+  return payments[0]?.data;
+}
+
+export async function getTradeData() {
+  const tradesBase = (filter = "") => `
+    SELECT count(*) as data 
+    FROM ${bigQueryEndpointBase}.history_trades
+    ${filter};
+  `;
+  const tradesQuery24h = tradesBase(`
+    WHERE ledger_closed_at >= TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL -24 HOUR) 
+    AND batch_run_date >= DATE_SUB(CURRENT_DATE(), INTERVAL 2 DAY)
+  `);
+  const tradesQuery48h = tradesBase(`
+    WHERE ledger_closed_at >= TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL -48 HOUR) 
```

Comment:
This could be a nit, but the comparison of past 24 hours to past 48 hours of trades is a little weird to me. I would have expected the comparison to be past 24 hours vs past 48-past 24 hours so that we're compared the same interval of trades. 

With the current definition, we should expect a minimum of 100% difference if we assume constant trading activity over time. @acharb do you know if this was how the old dashboard defined it?

Reply:
I'm currently using
```
AND batch_run_date <= DATE_SUB(CURRENT_DATE(), INTERVAL 1 DAY) 
AND batch_run_date >= DATE_SUB(CURRENT_DATE(), INTERVAL 2 DAY) 
```
for 48h, and 
```
AND batch_run_date >= DATE_SUB(CURRENT_DATE(), INTERVAL 1 DAY)
```
for 24hrs for both trades and volume queries, let me know if it's ideal

Reply:
> @acharb do you know if this was how the old dashboard defined it?

this is the first time we're showing Dex data on the dashboard.

Ya  I think this equation needs some adjusting, to compare the last 24h with the previous 24h:

```
const difference = results24h[0]?.data - results48h[0]?.data;
const output = (difference / results24h[0]?.data) * 100;
```
should instead be:
```
const previous24h =  results48h[0]?.data - results24h[0]?.data;
const output = (previous24h - results24h[0]?.data) / results24h[0]?.data * 100;
```

Reply:
Updated the equation on trades and dex data. I'll finish up testing and put up the commits in a few mins.

Reply:
Hey, using the updated equation i'm getting some really funky results. Let's say previous day = 42383, today = 75055, using this equation the result is 43%, but don't we want the percentage of increase/decrease? 

Using this as a calculator the results are very different: https://www.calculator.net/percent-calculator.html?c3par1=42383&c3par2=75055&ctype=3&x=86&y=23#pctdifference

Reply:
I think that's because on the denominator you should be using the previous 24 hr result, not current day. We should also flip numerator to find the difference between current day - previous 24hr:
```
(results24h[0]?.data  - previous24h) / previous24h * 100
```

Reply:
oops ya Syd's equation looks right, thanks for checking

Reply:
Hey, the results are still very weird, i'll give an example scenario to make it easier to talk about the equation since i'm terrible at math.
Let's say current day is 16, and the previous is 15. 
Day 15 results: 20
Day 16 results: 10
Expected result: 50% decrease
From what i understand the formula is the same, we only need to know if day 16 is smaller than day 15 or not to give a positive or negative result. So our code could be as such:

```
const increase =  Math.abs(sum48h - sum24h)
const difference = increase / sum48h * 100
const change = sum24h > sum48h ? difference : -Math.abs(difference)
```
Applying the previous scenario to this formula, the results would be:
```
const increase =  Math.abs(20 - 10)
const difference = 10 / 20 * 100
const change = 10 > 20 ? 50 : -50
```
This would output `-50`

that's correct right? With the following data i get +77% change.
![image](https://user-images.githubusercontent.com/8282304/170570282-88f28219-633d-4f29-82d5-8480e8a50d9e.png)

Reply:
this equation should be right: `(results24h - previous24h) / previous24h * 100`
(eg. from your example results24 = 10, previous24h = 20, gives -50%)

The dex-trades values don't look right. dex-trades48h should be greater than dex-trades24h, since it's the last 48 hours of trades vs the last 24 hours.

You can think of it this way (using the redis labels):
```
previous24h = dex-trades48h - dex-trades24h
results24h = dex-trades24h

change = (results24h - previous24h) / previous24h * 100
```

Reply:
Wait but our queries are not cumulative. The 48hrs query fetches from 2 days ago excluding the last 24hrs. Here's an example of filter used:
```
    AND batch_run_date <= DATE_SUB(CURRENT_DATE(), INTERVAL 1 DAY) 
    AND batch_run_date >= DATE_SUB(CURRENT_DATE(), INTERVAL 2 DAY) 
```

Reply:
ah sorry, I thought it was cumulative. So instead it would be:
```
previous24h = dex-trades48h
```
The % diff equation should be the same though

Reply:
are the numbers weird with that equation? I think let's use it for now, if we find the % diffs are weird we can take a look at it and adjust later

Reply:
Don't we have to check for negative values though? In this example, if the 48hrs query is larger than the 24hrs query, it would break. If we apply this equation in a decrease scenario, won't it break?

Using this equation
```
```

Reply:
That makes sense! I'm comitting the changes, here are the results on a volume query just to make sure we're on the same page:

48h: `75055.36712705049`
24h: `42382.75063436342`
change: `-77.08941964280275`

is that correct?

Reply:
@Selhar that looks good to me! The `change` is what I would expect it to be given those values.